iroh/src/node/rpc.rs

```diff
@@ -65,234 +66,235 @@ pub(crate) struct Handler<D> {
-                    })
-                    .await
-                }
-                DocList(msg) => {
-                    chan.server_streaming(msg, handler, |handler, req| {
-                        handler.inner.sync.doc_list(req)
-                    })
-                    .await
-                }
-                DocCreate(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_create(req).await
-                    })
-                    .await
-                }
-                DocDrop(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_drop(req).await
-                    })
-                    .await
-                }
-                DocImport(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_import(req).await
-                    })
-                    .await
-                }
-                DocSet(msg) => {
-                    let bao_store = handler.inner.db.clone();
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_set(&bao_store, req).await
-                    })
-                    .await
-                }
-                DocImportFile(msg) => {
-                    chan.server_streaming(msg, handler, Self::doc_import_file)
-                        .await
-                }
-                DocExportFile(msg) => {
-                    chan.server_streaming(msg, handler, Self::doc_export_file)
-                        .await
-                }
-                DocDel(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_del(req).await
-                    })
-                    .await
-                }
-                DocSetHash(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_set_hash(req).await
-                    })
-                    .await
-                }
-                DocGet(msg) => {
-                    chan.server_streaming(msg, handler, |handler, req| {
-                        handler.inner.sync.doc_get_many(req)
-                    })
-                    .await
-                }
-                DocGetExact(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_get_exact(req).await
-                    })
-                    .await
-                }
-                DocStartSync(msg) => {
-                    chan.rpc(msg, handler, |handler, req| async move {
-                        handler.inner.sync.doc_start_sync(req).await
-                    })
+        let handler = Self::new(inner);
+        join_set.spawn(async move {
```

Comment:
We are adding lots of tasks to the joinset. Presumably this is so they can all be properly shut down on shutdown.

But is it OK to add lots of tasks without ever calling join_next? won't you aggregate a lot of handles to long-finished short lived tasks?

Reply:
We do call join_next in the main loop, and break the main loop if a task panics - which is the main reason for the join set IMO (shutdown as well, but it won't matter much in practice because the tasks will also abort once the join set is dropped).